src/encoder.rs

```diff
@@ -297,12 +422,137 @@ impl<W: Write> Writer<W> {
             prev = line;
         }
         let zlib_encoded = zlib.finish()?;
-        for chunk in zlib_encoded.chunks(MAX_CHUNK_LEN as usize) {
-            self.write_chunk(chunk::IDAT, &chunk)?;
+        if self.sep_def_img || self.info.frame_control.is_none() {
+            self.sep_def_img = false;
+            for chunk in zlib_encoded.chunks(MAX_IDAT_CHUNK_LEN as usize) {
+                self.write_chunk(chunk::IDAT, &chunk)?;
+            }
+        } else if let Some(ref mut fctl) = self.info.frame_control {
+            fctl.encode(&mut self.w)?;
+            fctl.sequence_number += 1;
+
+            if self.written == 0 {
+                for chunk in zlib_encoded.chunks(MAX_IDAT_CHUNK_LEN as usize) {
+                    self.write_chunk(chunk::IDAT, &chunk)?;
+                }
+            } else {
+                let mut alldata = vec![0u8; 4 + MAX_fdAT_CHUNK_LEN as usize];
+                for chunk in zlib_encoded.chunks(MAX_fdAT_CHUNK_LEN as usize) {
+                    alldata[..4].copy_from_slice(&fctl.sequence_number.to_be_bytes());
+                    alldata[4..][..chunk.len()].copy_from_slice(chunk);
```

Comment:
That's a huge allocation even when it is not necessary. Selecting the minimum of the zlib_encoded's length and MAX_fdAT_CHUNK_LEN should do the trick and avoid this.

Reply:
Yup that's something I forgot to do, seems like I got confused by it at some point and added [..chunk.len()] where I should have done like you said

Reply:
Regarding this problem I think we could use the same `Limits` struct used for decoding, but this might be the subject of another PR